inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/sidebar/RecommendationSidebar.java

```diff
@@ -17,55 +17,241 @@
+				() -> recommendationService.getPreferences(aModel.getObject().getUser(),
+						aModel.getObject().getProject()),
+				(v) -> recommendationService.setPreferences(aModel.getObject().getUser(),
+						aModel.getObject().getProject(), v));
+
+		Form<Preferences> form = new Form<>("form", CompoundPropertyModel.of(model));
+
+		form.add(new NumberTextField<Integer>("maxPredictions", Integer.class).setMinimum(1).setMaximum(10).setStep(1));
+
+		form.add(new CheckBox("showAllPredictions"));
+
+		form.add(new LambdaAjaxButton<>("save", (_target, _form) -> aAnnotationPage.actionRefreshDocument(_target)));
+
+		add(form);
+
+		chartContainer = new Label("chart-container");
+		chartContainer.setOutputMarkupId(true);
+		add(chartContainer);
+	}
+
+	@Override
+	public void renderHead(IHeaderResponse aResponse) {
+		super.renderHead(aResponse);
+
+		// import Js
+		aResponse.render(JavaScriptHeaderItem.forReference(new WebjarsJavaScriptResourceReference("c3/current/c3.js")));
+		aResponse.render(
+				JavaScriptHeaderItem.forReference(new WebjarsJavaScriptResourceReference("d3js/current/d3.js")));
+
+		// import Css
+		aResponse.render(CssHeaderItem.forReference(new WebjarsCssResourceReference("c3/current/c3.css")));
+	}
+
+	@OnEvent
+	public void onRenderAnnotations(RenderAnnotationsEvent aEvent) {
+		try {
+			log.info("rendered annotation event");
+
+			StringBuilder chartType = new StringBuilder();
+			
+			HashMap<String, List<Double>> recommenderScoreMap = getLatestScores(aEvent);
+
+			// iterate over the recommender score map to create data arrays to feed to the
+			// c3 graph
+			StringBuilder dataColumns = new StringBuilder();
+			Iterator<Entry<String, List<Double>>> it = recommenderScoreMap.entrySet().iterator();
+			while (it.hasNext()) {
+				Entry<String, List<Double>> pair = it.next();
+
+				String jsonString = toJsonString(pair);
+				jsonString = jsonString.replaceAll("\".*?\"", "");
+				jsonString = jsonString.replaceAll("'.*?'", "");
+				jsonString = jsonString.replaceAll("`.*?`", "");
+
+				String data = substring(jsonString, 3, -2);
+
+				// append recommender name
+				dataColumns.append("['");
+				String[] recommenderClass = pair.getKey().toString().split("\\.");
+				String recommenderName = recommenderClass[recommenderClass.length - 1];
+				
+				//define chart type for the recommender
+				chartType.append(recommenderName);
+				chartType.append(": 'step', ");
+				
+				dataColumns.append(recommenderName);
+
+				// append data columns
+				dataColumns.append("', ");
+				dataColumns.append(data);
+				dataColumns.append("]");
+				dataColumns.append(",");
+
+				// avoids a ConcurrentModificationException
+				it.remove();
+			}
+
+			String javascript = createJSScript(dataColumns.toString(), chartType.toString());
+			log.info(javascript);
+
+			aEvent.getRequestHandler().prependJavaScript(javascript);
+		} catch (IOException e) {
+			log.error("Unable to render chart", e);
```

Comment:
Where exactly is the error thrown? This try-catch looks really long, I would try to make it as short as possible.